File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/dao/DataRepository.java
package uk.gov.justice.digital.hmpps.riskprofiler.dao;

import lombok.extern.slf4j.Slf4j;
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.RiskDataSet;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.regex.Pattern;

@Slf4j
public abstract class DataRepository<F extends RiskDataSet> {

    final Pattern NOMS_ID_REGEX = Pattern.compile("^[A-Z]\\d{4}[A-Z]{2}$");

    private final ImportedFile<F> dataA = new ImportedFile<>();
    private final ImportedFile<F> dataB = new ImportedFile<>();
    private volatile AtomicBoolean isA = new AtomicBoolean(true);

    public void process(List<List<String>> csvData, String filename, LocalDateTime timestamp) {
        doProcess(csvData,  filename,  timestamp, getStandbyData());
        toggleData();
    }

    protected abstract void doProcess(List<List<String>> csvData, String filename, LocalDateTime timestamp, ImportedFile<F> data);

    public Optional<F> getByKey(String key) {
        if (getData().getDataSet() != null) {
            return Optional.ofNullable(getData().getDataSet().get(key));
        }
        return Optional.empty();
    }

    public LocalDateTime getFileTimestamp() {
        return (isA.get() ? dataA : dataB).getFileTimestamp();
    }

    public ImportedFile<F> getData() {
        return isA.get() ? dataA : dataB;
    }

    public ImportedFile<F> getStandbyData() {
        return isA.get() ? dataB : dataA;
    }

    private void toggleData() {
        isA.set(!isA.get());
        log.debug("Switched to {} data map {}", getData().getFileType(), isA.get() ? "A" : "B");
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/dao/DataRepository.kt
package uk.gov.justice.digital.hmpps.riskprofiler.dao

import org.slf4j.LoggerFactory
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.RiskDataSet
import java.time.LocalDateTime
import java.util.Optional
import java.util.Optional.empty
import java.util.Optional.ofNullable
import java.util.concurrent.atomic.AtomicBoolean
import java.util.regex.Pattern

abstract class DataRepository<F : RiskDataSet> {
  @JvmField
  val NOMS_ID_REGEX = Pattern.compile("^[A-Z]\\d{4}[A-Z]{2}$")
  internal open val dataA = ImportedFile<F>()
  internal open val dataB = ImportedFile<F>()
  internal open val isA = AtomicBoolean(true)
  internal open val dataAvailable = AtomicBoolean(false)
  fun process(csvData: List<List<String>>, filename: String, timestamp: LocalDateTime) {
    doProcess(csvData, filename, timestamp, standbyData)
    toggleData()
    dataAvailable.set(true)
  }

  protected abstract fun doProcess(
    csvData: List<List<String>>,
    filename: String,
    timestamp: LocalDateTime,
    data: ImportedFile<F>
  )

  open fun getByKey(key: String?): Optional<F> {
    return if (data.dataSet != null) {
      Optional.ofNullable(data.dataSet!![key!!])
    } else {
      Optional.empty()
    }
  }

  val fileTimestamp: LocalDateTime?
    get() = (if (isA.get()) dataA else dataB).fileTimestamp
  val data: ImportedFile<F>
    get() = if (isA.get()) dataA else dataB
  val standbyData: ImportedFile<F>
    get() = if (isA.get()) dataB else dataA

  private fun toggleData() {
    isA.set(!isA.get())
    log.debug("Switched to {} data map {}", data.fileType, if (isA.get()) "A" else "B")
  }

  fun dataAvailable(): Boolean = dataAvailable.get()

  companion object {
    private val log = LoggerFactory.getLogger(DataRepository::class.java)
  }
}
